kong/router.lua

```diff
@@ -1533,76 +1673,118 @@ function _M.new(routes)
 
     -- header match
 
-    for _, header_name in ipairs(plain_indexes.headers) do
-      if req_headers[header_name] then
-        req_category = bor(req_category, MATCH_RULES.HEADER)
-        hits.header_name = header_name
-        break
+    local headers_key do
+      local headers_count = 0
+      if match_headers then
+        for i = 1, plain_indexes.headers[0] do
+          local name = plain_indexes.headers[i]
+          local value = req_headers[name]
+          if value then
+            if type(value) == "table" then
+              value = clone(value)
+              for i = 1, #value do
+                value[i] = lower(value[i])
+              end
+              sort(value)
+              value = concat(value, ", ")
+
+            else
+              value = lower(value)
+            end
+
+            if headers_count == 0 then
+              headers_key = { "|", name, "=", value }
+
+            else
+              headers_key[headers_count+1] = "|"
+              headers_key[headers_count+2] = name
+              headers_key[headers_count+3] = "="
+              headers_key[headers_count+4] = value
+            end
+
+            headers_count = headers_count + 4
+
+            if not hits.header_name then
+              hits.header_name = name
+              req_category = bor(req_category, MATCH_RULES.HEADER)
+            end
```

Comment:
What does this code do?

Reply:
I think there is some kind of optimization for single header hits. So it is stored there. It will still match others too. But as headers are matched with `OR` semantics, it just reduces possible routes by single header. And yes, it marks that there are routes for this request that need header matching.